src/cdi_CPEloss/Tabular_CP_Eloss.cc

```diff
@@ -0,0 +1,270 @@
+      read_line();                                // Z, A, mass
+      if (zaid_target_ion == target.get_zaid()) {
+        // This is the requested target ion
+        target_found = true;
+        uint32_t nentry = 0;
+        for (uint32_t n = 0; n < nlines; n++) {
+          line_entries = read_line();
+          for (std::string entry : line_entries) {
+            stopping_data_1d[nentry] = stod(entry);
+            nentry++;
+          }
+        }
+        break;
+      } else {
+        // This is not the requested target ion
+        for (uint32_t n = 0; n < nlines; n++) {
+          read_line();
+        }
+      }
+    }
+  }
+  file.close();
+
+  std::ostringstream dedx_error;
+  dedx_error << "Error finding target ZAID \"" << target.get_zaid()
+             << "\" in DEDX file \"" << filename << "\"";
+  Insist(target_found, dedx_error.str());
+
+  stopping_data.resize(n_energy, n_density, n_temperature);
+  for (uint32_t ne = 0; ne < n_energy; ne++) {
+    for (uint32_t nd = 0; nd < n_density; nd++) {
+      for (uint32_t nt = 0; nt < n_temperature; nt++) {
+        stopping_data(ne, nd, nt) =
+            stopping_data_1d[ne + n_energy * (nd + n_density * nt)];
+      }
+    }
+  }
+
+  // Convert units on table to match those of getEloss:
+  //   energy:      MeV -> cm/shk (using target particle mass)
+  double energy_cgs = exp(min_log_energy) * (1.e6 * 1.6021772e-12);
+  min_log_energy = log(sqrt(2. * energy_cgs / target.get_mass()) * 1.e-8);
+  d_log_energy = d_log_energy / 2.;
+  //   density:     cm^-3 -> g cm^-3
+  min_log_density = log(exp(min_log_density) * target.get_mass());
+  //   temperature: keV -> keV
+  // Note that d log x = dx / x is not affected by unit conversion factors
+  for (auto &energy : energies) {
+    energy =
+        sqrt(2. * (energy * 1.e6 * 1.6021772e-12) / target.get_mass()) * 1.e-8;
+  }
+  for (auto &density : densities) {
+    density *= target.get_mass();
+  }
+
+  // Initialize table bounds
+  min_energy = exp(min_log_energy);
+  max_energy = exp(min_log_energy + d_log_energy * (n_energy - 1));
+  min_density = exp(min_log_density);
+  max_density = exp(min_log_density + d_log_density * (n_density - 1));
+  min_temperature = exp(min_log_temperature);
+  max_temperature =
+      exp(min_log_temperature + d_log_temperature * (n_temperature - 1));
+}
+
+/*!
+ * \brief Read a line from an eloss datafile and return as a vector of strings.
+ *
+ * Convenience function, especially when reading header data which is not
+ * uniform in the number of entries per line, or the types of those entries.
+ *
+ * \return entries the resulting vector of entries in the datafile line.
+ *
+ */
+std::vector<std::string> Tabular_CP_Eloss::read_line() {
+  std::string line;
+  std::getline(file, line);
+  std::vector<std::string> entries;
+  std::istringstream iss(line);
+  for (std::string s; iss >> s;) {
+    entries.push_back(s);
+  }
+  return entries;
+}
```

Comment:
@keadyk Do you have any thoughts about making this function less verbose?

Reply:
Hmm, my initial thought was:

http://www.cplusplus.com/reference/cstring/strtok/

but this is an old c-style string function, so I don't think it makes sense to hack it in here. Honestly, I'm not sure there's a more elegant solution to tokenizing std::strings!

Reply:
In `ds++/DracoStrings.hh`, I have a function `tokenize` that converts a string into a vector of words, but it is more code because it allows any character to be used as a delimiter (not just a space).

` std::vector<std::string> entries = rtt_dsxx::tokenize(line);`

Reply:
Oh right!! I should've remembered that, because I've used it before :)